src/HttpRequest.h

```diff
@@ -193,6 +193,9 @@ class HttpRequest: public Http::Message
     /// whether we have responded with HTTP 100 or FTP 150 already
     bool forcedBodyContinuation;
 
+    /// whether we are currently creating a CONNECT header (to be sent to peer)
+    bool generatingConnect = false;
```

Comment:
> PR description: It is a flag to control ACL matching against the HttpRequest data

> Alex: What makes you think that? AFACIT, commit [090f1d3](https://github.com/squid-cache/squid/commit/090f1d3c3c4e1aa9ab96b84787907967313ccfcf) reached (and painstakingly documented) essentially the opposite conclusion.

> Amos: The actual current code usage of this flag.

That answer does not compute for me because the "actual current code usage of this flag" does not "control ACL matching against the HttpRequest data". The current code uses this flag to determine whether the current code context is the `CONNECT` request generation context. That generation uses many objects, including HttpRequest and MasterXaction. More specifically, the `at_step` ACL matches the current master transaction processing step, not "HttpRequest data". The transaction processing step concept lives on a level higher than HttpRequest.

> It should really be a flag of HttpStateData object, but that object is not passed to the access control code.

That would not be a good idea because `CONNECT` generation is not always related to or uses HttpStateData objects.

> The HttpRequest object used by HttpStateData::httpBuildRequestHeader() is where those access controls get details from HttpStateData.

The primary ACL  relevant here -- `at_step` -- gets the current step information from ConnStateData and MasterXaction objects. The ConnStateData source is a temporary flaw -- that source should be MasterXaction or equivalent instead. The MasterXaction source is not perfect either, but it is the best we can do for now AFAICT.

In summary, the key assumption behind this PR remains unsubstantiated and contradicts commit 090f1d3 conclusions. I believe it is also false, but there is another change request dedicated to that problem.

Reply:
> > PR description: It is a flag to control ACL matching against the HttpRequest data
> 
> > Alex: What makes you think that? AFACIT, commit [090f1d3](https://github.com/squid-cache/squid/commit/090f1d3c3c4e1aa9ab96b84787907967313ccfcf) reached (and painstakingly documented) essentially the opposite conclusion.
> 
> > Amos: The actual current code usage of this flag.
> 
> Alex:  That answer does not compute for me because the "actual current code usage of this flag" does not "control ACL matching against the HttpRequest data".
> The current code uses this flag to determine whether the current code context is the `CONNECT` request generation context.


> That generation uses many objects, including HttpRequest and MasterXaction

Specifically it uses HttpRequest as the primary object outside ACL namespace to locate this flag.
 ```
-    return checklist->request->masterXaction->generatingConnect ? 1 : 0;
+    return checklist->request->generatingConnect ? 1 : 0;
```

> . More specifically, the `at_step` ACL matches the current master transaction processing step, not "HttpRequest data".

You are comparing different things. Abstract description of **entire** ACL purpose naturally is not the same as where the implementation gets data to check.

Also, as I am sure you know one piece of "HttpRequest data" is the up-reference to which "master transaction" the HttpRequest is a part of.

> The transaction processing step concept lives on a level higher than HttpRequest.
> 

Indeed, abstract concept does exist at a higher level than any one of the components implementing that concept.

Please be aware that there are at least 2 layers of transaction nested between the HttpStateData context which is performing the "current transaction step" and the "master transaction" context by any reasonable definition of "master" and "transaction". The "master transaction" several layers earlier may have nothing to do with this need to generate a CONNECT tunnel.

By the time this code occurs there are potentially 6 HttpRequest objects in play representing the different layers of transaction nesting and adaptation sub-transactions (all of which are "master transaction", hmm). This flag indicates the one being used by ACL as data source (checklist->request) represents a transaction generating a new HttpRequest with CONNECT method.

It is wrong to set in MasterXaction because ACL checks are not necessarily synchronous and this flag should not affect any of the other transaction HttpRequest objects linked to the MasterXaction.

AFAICT this code only works today because the ACLs and all HttpStateData methods using the flag are completely synchronous for the entire period it is set to true.


> > It should really be a flag of HttpStateData object, but that object is not passed to the access control code.
> 
> That would not be a good idea because `CONNECT` generation is not always related to or uses HttpStateData objects.

This "CONNECT request generation context" is part of the HttpStateData internal logics. So yes it is always **directly** related to and **requires** an HttpStateData object.

> 
> > The HttpRequest object used by HttpStateData::httpBuildRequestHeader() is where those access controls get details from HttpStateData.
> 
> The primary ACL relevant here -- `at_step` -- gets the current step information from ConnStateData and MasterXaction objects. The ConnStateData source is a temporary flaw -- that source should be MasterXaction or equivalent instead. The MasterXaction source is not perfect either, but it is the best we can do for now AFAICT.
> 

No. ```return checklist->request->....generatingConnect ? 1 : 0;```

HttpRequest is better than MasterXaction as demonstrated by this PR code simplification and reducing side effects of the flag being set.

> In summary, the key assumption behind this PR

Which unstated assumption are you refering to?